### PR TITLE
Add rebate per reserve type

### PR DESCRIPTION
**Pull request by Anyhowclick**

### CODE REVIEW

File: contractsSol5/IKyberStorage.sol [hunk 1]
```diff
@@ -73,7 +73,8 @@ contract IKyberStorage {
         returns (
             bytes32 reserveId,
             ReserveType resType,
-            bool isFeeAccountedFlags
+            bool isFeeAccounted,
```
Comment (ilanDoron) [inline contractsSol5/IKyberStorage.sol:76]: please have flags in it
for both.

was discussed on our chat with Victor and Mike.

Idea is when we have values for a few reserves. its plural.
and here flags make sense.
or u have better one?

File: contractsSol5/KyberNetwork.sol [hunk 2]
```diff
@@ -1137,7 +1145,11 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils4, IKyberNetwork, Reentra
 
             if (reservesData.isFeeAccountedFlags[i]) {
                 tradeData.feeAccountedBps += reservesData.splitsBps[i];
-                tradeData.numFeeAccountedReserves++;
+            }
+
+            if (reservesData.isEntitledRebateFlags[i]) {
```
Comment (ilanDoron) [inline contractsSol5/KyberNetwork.sol:1150]: this if should be inside isFeeAccountedFlags

File: contractsSol5/KyberStorage.sol [hunk 3]
```diff
@@ -214,6 +216,27 @@ contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
...
+        uint256 entitledRebateData;
+
+        if (apr) entitledRebateData |= 1 << uint256(ReserveType.APR);
+        if (fpr) entitledRebateData |= 1 << uint256(ReserveType.FPR);
+        if (bridge) entitledRebateData |= 1 << uint256(ReserveType.BRIDGE);
+        if (utility) entitledRebateData |= 1 << uint256(ReserveType.UTILITY);
+        if (custom) entitledRebateData |= 1 << uint256(ReserveType.CUSTOM);
+        if (orderbook) entitledRebateData |= 1 << uint256(ReserveType.ORDERBOOK);
+
+        entitledRebatePerType = entitledRebateData;
```
Comment (ilanDoron) [inline contractsSol5/KyberStorage.sol:237]: we can consider that this should always be updated after isFeePaying is updated.
and then must be aligned.
can't have rebate if not feeAccounted.

File: contractsSol5/KyberStorage.sol [hunk 4]
```diff
@@ -299,37 +322,43 @@ contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
...
             ReserveType resType,
-            bool isFeeAccountedFlags
+            bool isFeeAccounted,
+            bool isEntitledRebate
         )
     {
         reserveAddress = reserveIdToAddresses[reserveId][0];
         resType = ReserveType(reserveType[reserveId]);
-        isFeeAccountedFlags = (feeAccountedPerType & (1 << reserveType[reserveId])) > 0;
+        isFeeAccounted = (feeAccountedPerType & (1 << reserveType[reserveId])) > 0;
```
Comment (ilanDoron) [inline contractsSol5/KyberStorage.sol:339]: here and below we should use resType
and not call storage

File: contractsSol5/KyberStorage.sol [hunk 5]
```diff
@@ -299,37 +322,43 @@ contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
...
             ReserveType resType,
-            bool isFeeAccountedFlags
+            bool isFeeAccounted,
+            bool isEntitledRebate
         )
     {
         reserveId = reserveAddressToId[reserve];
         resType = ReserveType(reserveType[reserveId]);
-        isFeeAccountedFlags = (feeAccountedPerType & (1 << reserveType[reserveId])) > 0;
+        isFeeAccounted = (feeAccountedPerType & (1 << reserveType[reserveId])) > 0;
```
Comment (ilanDoron) [inline contractsSol5/KyberStorage.sol:360]: use resType

File: test/sol5/kyberMatchingEngine.js [hunk 6]
```diff
@@ -526,278 +528,6 @@ contract('KyberMatchingEngine', function(accounts) {
     });
 });
 
-async function fetchReservesRatesFromRateHelper(matchingEngineInstance, rateHelperInstance, reserveInstances, tokenAddress, qty, networkFeeBps, isTokenToEth) {
```
Comment (ilanDoron) [inline test/sol5/kyberMatchingEngine.js:529]: are all these deletions related to this PR?

File: contractsSol5/IKyberStorage.sol (hunk 1 above)
Comment (Anyhowclick) [reply contractsSol5/IKyberStorage.sol:76]: As discussed, isFeeAccountedFlag

File: contractsSol5/KyberStorage.sol (hunk 3 above)
Comment (Anyhowclick) [reply contractsSol5/KyberStorage.sol:244]: see updated code

File: contractsSol5/KyberNetwork.sol [hunk 7]
```diff
@@ -885,10 +888,12 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils4, IKyberNetwork, Reentra
                 (tradeData.input.dest != ETH_TOKEN_ADDRESS),
             hint
         );
-        reservesData.isFeeAccountedFlags = kyberStorage.getFeeAccountedData(reservesData.ids);
+        (reservesData.isFeeAccountedFlags, reservesData.isEntitledRebateFlags) = kyberStorage
```
Comment (ilanDoron) [inline contractsSol5/KyberNetwork.sol:891]: can u fix ugly prettier?

File: contractsSol5/KyberNetwork.sol [hunk 8]
```diff
@@ -1137,7 +1145,11 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils4, IKyberNetwork, Reentra
 
             if (reservesData.isFeeAccountedFlags[i]) {
                 tradeData.feeAccountedBps += reservesData.splitsBps[i];
-                tradeData.numFeeAccountedReserves++;
+
+                if (reservesData.isEntitledRebateFlags[i]) {
+                tradeData.entitledRebateBps += reservesData.splitsBps[i];
```
Comment (ilanDoron) [inline contractsSol5/KyberNetwork.sol:1150]: indent two lines

File: contractsSol5/KyberStorage.sol (hunk 3 above)
Comment (ilanDoron) [reply contractsSol5/KyberStorage.sol:261]: Yes
Great

File: contractsSol5/KyberNetwork.sol (hunk 7 above)
Comment (Anyhowclick) [reply contractsSol5/KyberNetwork.sol:891]: Not sure what's the ideal style here... see if it's ok in update

**Pull request merged**